File: apps/desktop/src/renderer/services/workflows.ts
/**
 * User workflows REST client (定案 §10.6 / §10.7 / §10.8).
 *
 * Wire shapes hand-written (camelCase), aligned with OpenAPI; do not gen:types.
 * Official templates: GET /v1/workflow-playbook-templates；复制:
 * POST /v1/workflows/from-playbook. Templates 404/501 → empty list (UI hides section).
 * Slot definitions come from the backend catalog only — no local replica.
 *
 * 两处「槽位」不是一回事：`WorkflowTemplateSlot` 是官方模板复制时要填的参数目录；
 * definition 顶层的 `WorkflowSlot`（见 workflowDefinition.ts）是已存工作流里可换值的
 * 参数，跑一次时用它的 `default` 预填。
 */

import { ApiError, api } from "@/services/api";
import {
  type WorkflowDefinition,
  emptyWorkflowDefinition,
  parseWorkflowDefinition,
} from "@/services/workflowDefinition";
import {
  type WorkflowSource,
  parseWorkflowSource,
} from "@/services/workflowSource";

/** Built-in schedule presets (UI + PUT trigger). Custom uses `cron`. */
export type SchedulePreset =
  | "daily"
  | "weekdays"
  | "weekly_mon"
  | "weekly_fri"
  | "monthly_1"
  | "custom";

export type TriggerKind = "schedule" | "webhook";

export const SCHEDULE_PRESET_ORDER: SchedulePreset[] = [
  "daily",
  "weekdays",
  "weekly_mon",
  "weekly_fri",
  "monthly_1",
  "custom",
];

export const SCHEDULE_PRESET_LABELS: Record<SchedulePreset, string> = {
  daily: "每天",
  weekdays: "工作日",
  weekly_mon: "每周一",
  weekly_fri: "每周五",
  monthly_1: "每月 1 日",
  custom: "自定义 cron",
};

export const TRIGGER_KIND_ORDER: TriggerKind[] = ["schedule", "webhook"];

export const TRIGGER_KIND_LABELS: Record<TriggerKind, string> = {
  schedule: "定时",
  webhook: "Webhook",
};

export interface WorkflowTrigger {
  kind: TriggerKind;
  schedulePreset: SchedulePreset | null;
  cron: string | null;
  folderId: string;
  enabled: boolean;
  nextRunAt: string | null;
  lastRunAt: string | null;
  lastError: string | null;
  webhookId: string | null;
  /** Relative path, e.g. `/v1/hooks/workflows/{webhook_id}`. */
  webhookUrl: string | null;
  /** One-shot plaintext; only on PUT / rotate responses. */
  webhookSecret: string | null;
}

export interface PutWorkflowTriggerInput {
  kind: TriggerKind;
  folderId: string;
  enabled?: boolean;
  schedulePreset?: SchedulePreset;
  cron?: string | null;
}

export interface RotateWorkflowSecretResult {
  webhookSecret: string;
  webhookUrl: string | null;
  webhookId: string | null;
}

export interface UserWorkflow {
  id: string;
  name: string;
  description: string | null;
  definition: WorkflowDefinition;
  /** 出处（服务端权威字段，见 workflowSource.ts）；`null` = 不是固化来的。 */
  source: WorkflowSource | null;
  /** Clock / webhook; `null` / omitted = 只手点「跑一次」。 */
  trigger?: WorkflowTrigger | null;
  version: number;
  createdAt: string;
  updatedAt: string;
}

export interface CreateWorkflowInput {
  name: string;
  description?: string | null;
  definition?: WorkflowDefinition;
}

export interface PatchWorkflowInput {
  name?: string;
  description?: string | null;
  definition?: WorkflowDefinition;
}

export interface RunWorkflowInput {
  folderId: string;
  /** Optional per-run supplement (本轮补充). */
  note?: string | null;
  /**
   * 本轮槽位覆盖（key → 值）。只带用户改过的槽位：没带的 key 由服务端回落到
   * definition 里的 `default`，所以一个都不带 = 按固化那轮原样重跑。
   */
  slots?: Record<string, string>;
}

export interface RunWorkflowResult {
  conversationId: string | null;
}

/** One allowed value of an enumerated slot (render a picker, not a textbox). */
export interface WorkflowTemplateSlotChoice {
  value: string;
  label: string;
}

/** Official playbook template slot (domain) — mirrors the backend catalog. */
export interface WorkflowTemplateSlot {
  key: string;
  label: string;
  required: boolean;
  hint: string | null;
  /** Non-empty → only these values are accepted by the backend. */
  choices: WorkflowTemplateSlotChoice[];
}

/** Official playbook → copy-as-mine catalog entry (定案 §10.8). */
export interface WorkflowTemplate {
  id: string;
  title: string;
  summary: string;
  slots: WorkflowTemplateSlot[];
}

export interface FromPlaybookInput {
  playbook: string;
  /** Optional display name for the new user workflow. */
  name?: string | null;
  /** Primary slot values (topic / feature / task / app …). */
  slots: Record<string, string>;
}

/** Wire: snake_case — mirrors OpenAPI. */
/** Wire: nested `trigger` on GET/PATCH WorkflowSummary. */
export interface WorkflowTriggerWire {
  kind?: string | null;
  trigger_kind?: string | null;
  schedule_preset?: string | null;
  cron?: string | null;
  folder_id?: string | null;
  enabled?: boolean;
  next_run_at?: string | null;
  last_run_at?: string | null;
  last_error?: string | null;
  webhook_id?: string | null;
  webhook_url?: string | null;
  webhook_secret?: string | null;
}

export interface UserWorkflowWire {
  id: string;
  name: string;
  description?: string | null;
  definition: unknown;
  /** `WorkflowSummary` 顶层的出处；definition 里的同名键不再是它。 */
  source?: unknown;
  trigger?: WorkflowTriggerWire | null;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface RotateWorkflowSecretWire {
  webhook_secret?: string | null;
  webhook_url?: string | null;
  webhook_id?: string | null;
}

/** Wire: GET /v1/workflow-playbook-templates item. */
export interface WorkflowTemplateSlotChoiceWire {
  value?: string;
  label?: string;
}

export interface WorkflowTemplateSlotWire {
  key: string;
  label?: string;
  required?: boolean;
  hint?: string | null;
  choices?: WorkflowTemplateSlotChoiceWire[];
}

export interface WorkflowTemplateWire {
  id: string;
  title?: string;
  summary?: string;
  /** Prose one-liner for help copy; `slots` is the machine-readable source. */
  primary_slots?: string;
  slots?: WorkflowTemplateSlotWire[];
}

function asPreset(raw: string | null | undefined): SchedulePreset | null {
  if (!raw) return null;
  return (SCHEDULE_PRESET_ORDER as string[]).includes(raw)
    ? (raw as SchedulePreset)
    : "custom";
}

function asTriggerKind(raw: string | null | undefined): TriggerKind {
  return raw === "webhook" ? "webhook" : "schedule";
}

/** Display / copy as a relative path; never absolutize. */
export function relativeWebhookUrl(
  raw: string | null | undefined,
  webhookId?: string | null,
): string | null {
  if (raw) {
    if (/^https?:\/\//i.test(raw)) {
      try {
        return new URL(raw).pathname || null;
      } catch {
        const path = raw.replace(/^https?:\/\/[^/]+/i, "");
        return path.startsWith("/") ? path : `/${path}`;
      }
    }
    return raw.startsWith("/") ? raw : `/${raw}`;
  }
  const id = webhookId?.trim();
  if (!id) return null;
  return `/v1/hooks/workflows/${encodeURIComponent(id)}`;
}

export function toWorkflowTrigger(
  raw: WorkflowTriggerWire | null | undefined,
): WorkflowTrigger | null {
  if (!raw || typeof raw !== "object") return null;
  const kind = asTriggerKind(raw.kind ?? raw.trigger_kind);
  const webhookId = raw.webhook_id ?? null;
  const lastError = raw.last_error?.trim() || null;
  return {
    kind,
    schedulePreset: asPreset(raw.schedule_preset),
    cron: raw.cron ?? null,
    folderId: raw.folder_id ?? "",
    enabled: raw.enabled !== false,
    nextRunAt: raw.next_run_at ?? null,
    lastRunAt: raw.last_run_at ?? null,
    lastError,
    webhookId,
    webhookUrl:
      kind === "webhook"
        ? relativeWebhookUrl(raw.webhook_url, webhookId)
        : null,
    webhookSecret: raw.webhook_secret ?? null,
  };
}

export function toUserWorkflow(w: UserWorkflowWire): UserWorkflow {
  return {
    id: w.id,
    name: w.name,
    description: w.description ?? null,
    definition: parseWorkflowDefinition(w.definition),
    source: parseWorkflowSource(w.source),
    trigger: toWorkflowTrigger(w.trigger),
    version: w.version,
    createdAt: w.created_at,
    updatedAt: w.updated_at,
  };
}

/**
 * Parse UTC daily cron ``M H * * *`` → local wall-clock hour/minute.
 * Falls back to 09:00 local when the expression is missing or not daily.
 */
export function localHmFromUtcCron(cron: string | null | undefined): {
  hour: number;
  minute: number;
} {
  const m = cron?.trim().match(/^(\d{1,2})\s+(\d{1,2})\s+\*\s+\*\s+\*$/);
  if (!m) return { hour: 9, minute: 0 };
  const utcMin = Number(m[1]);
  const utcHour = Number(m[2]);
  if (
    !Number.isFinite(utcMin) ||
    !Number.isFinite(utcHour) ||
    utcMin < 0 ||
    utcMin > 59 ||
    utcHour < 0 ||
    utcHour > 23
  ) {
    return { hour: 9, minute: 0 };
  }
  const d = new Date();
  d.setUTCHours(utcHour, utcMin, 0, 0);
  return { hour: d.getHours(), minute: d.getMinutes() };
}

/** Local hour/minute → UTC daily cron ``M H * * *``. */
export function utcCronFromLocalHm(hour: number, minute: number): string {
  const h = Math.max(0, Math.min(23, Math.floor(hour)));
  const min = Math.max(0, Math.min(59, Math.floor(minute)));
  const d = new Date();
  d.setHours(h, min, 0, 0);
  return `${d.getUTCMinutes()} ${d.getUTCHours()} * * *`;
}

/** Row subtitle for an attached trigger. */
export function triggerSummary(
  trigger: WorkflowTrigger | null | undefined,
): string | null {
  if (!trigger) return null;
  if (trigger.kind === "webhook") return TRIGGER_KIND_LABELS.webhook;
  if (trigger.schedulePreset && trigger.schedulePreset !== "custom") {
    return SCHEDULE_PRESET_LABELS[trigger.schedulePreset];
  }
  if (trigger.cron) return trigger.cron;
  if (trigger.schedulePreset === "custom") return "自定义";
  return TRIGGER_KIND_LABELS.schedule;
}

function toTemplateSlotChoice(
  raw: WorkflowTemplateSlotChoiceWire,
): WorkflowTemplateSlotChoice {
  const value = String(raw.value ?? "").trim();
  return { value, label: raw.label?.trim() || value };
}

function toTemplateSlot(raw: WorkflowTemplateSlotWire): WorkflowTemplateSlot {
  const key = String(raw.key ?? "").trim();
  return {
    key,
    label: raw.label?.trim() || key,
    required: raw.required === true,
    hint: raw.hint?.trim() || null,
    choices: (raw.choices ?? [])
      .map(toTemplateSlotChoice)
      .filter((c) => c.value),
  };
}

export function toWorkflowTemplate(w: WorkflowTemplateWire): WorkflowTemplate {
  const id = String(w.id ?? "").trim();
  return {
    id,
    title: w.title?.trim() || id,
    summary: w.summary?.trim() ?? "",
    slots: (w.slots ?? []).map(toTemplateSlot).filter((s) => s.key),
  };
}

function isMissingRoute(e: unknown): boolean {
  return e instanceof ApiError && (e.status === 404 || e.status === 501);
}

/** List the signed-in user's workflows. */
export async function listWorkflows(): Promise<UserWorkflow[]> {
  const res = await api.get<UserWorkflowWire[]>("/v1/workflows");
  return (Array.isArray(res) ? res : []).map(toUserWorkflow);
}

export async function getWorkflow(id: string): Promise<UserWorkflow> {
  const res = await api.get<UserWorkflowWire>(
    `/v1/workflows/${encodeURIComponent(id)}`,
  );
  return toUserWorkflow(res);
}

export async function createWorkflow(
  input: CreateWorkflowInput,
): Promise<UserWorkflow> {
  const res = await api.post<UserWorkflowWire>("/v1/workflows", {
    name: input.name.trim(),
    description: input.description?.trim() || null,
    definition: input.definition ?? emptyWorkflowDefinition(),
  });
  return toUserWorkflow(res);
}

export async function patchWorkflow(
  id: string,
  input: PatchWorkflowInput,
): Promise<UserWorkflow> {
  const body: Record<string, unknown> = {};
  if (input.name !== undefined) body.name = input.name.trim();
  if (input.description !== undefined) {
    const trimmed = input.description?.trim() || "";
    if (!trimmed) {
      body.clear_description = true;
    } else {
      body.description = trimmed;
    }
  }
  if (input.definition !== undefined) body.definition = input.definition;
  const res = await api.patch<UserWorkflowWire>(
    `/v1/workflows/${encodeURIComponent(id)}`,
    body,
  );
  return toUserWorkflow(res);
}

export async function deleteWorkflow(id: string): Promise<void> {
  await api.delete(`/v1/workflows/${encodeURIComponent(id)}`);
}

/** Run once: select workspace + optional note (+ slot overrides) → direct-start bypass. */
export async function runWorkflow(
  id: string,
  input: RunWorkflowInput,
): Promise<RunWorkflowResult> {
  const body: Record<string, unknown> = {
    folder_id: input.folderId,
    note: input.note?.trim() || null,
  };
  const slots: Record<string, string> = {};
  for (const [k, v] of Object.entries(input.slots ?? {})) {
    const key = k.trim();
    const val = v.trim();
    if (key && val) slots[key] = val;
  }
  // 空覆盖不进 body：请求与「没有槽位的工作流」逐字一致。
  if (Object.keys(slots).length > 0) body.slots = slots;
  const res = await api.post<{ conversation_id?: string | null }>(
    `/v1/workflows/${encodeURIComponent(id)}/run`,
    body,
  );
  return { conversationId: res?.conversation_id ?? null };
}

/**
 * 按需抽出可换参数：返回这条工作流抽完后的最新形态（definition 里带上 slots，
 * 任务文本里的原值换成了 `{{key}}` 占位符）。
 *
 * 抽不出来时返回的 definition 与调用前逐字一致（仍然没有 slots）——这不是错误，
 * 照常直接跑即可。已有槽位时服务端幂等直接返回。抽取要真跑一次模型（最长约 20 秒），
 * 调用方须把它放在不挡用户的路径上。
 */
export async function suggestWorkflowSlots(id: string): Promise<UserWorkflow> {
  const res = await api.post<UserWorkflowWire>(
    `/v1/workflows/${encodeURIComponent(id)}/suggest-slots`,
  );
  return toUserWorkflow(res);
}

function putTriggerBody(
  input: PutWorkflowTriggerInput,
): Record<string, unknown> {
  const kind = input.kind;
  const body: Record<string, unknown> = {
    kind,
    folder_id: input.folderId,
    enabled: input.enabled ?? true,
  };
  if (kind === "schedule") {
    const preset = input.schedulePreset ?? "weekly_mon";
    body.schedule_preset = preset;
    if (preset === "custom") {
      body.cron = input.cron ?? null;
    }
  }
  return body;
}

/** Attach or replace the clock / webhook. Response is the workflow with `trigger`. */
export async function putWorkflowTrigger(
  id: string,
  input: PutWorkflowTriggerInput,
): Promise<UserWorkflow> {
  const res = await api.put<UserWorkflowWire>(
    `/v1/workflows/${encodeURIComponent(id)}/trigger`,
    putTriggerBody(input),
  );
  return toUserWorkflow(res);
}

/** Clear the trigger. Prefers a workflow payload; otherwise re-GETs. */
export async function deleteWorkflowTrigger(id: string): Promise<UserWorkflow> {
  const res = await api.delete<UserWorkflowWire | { status?: string }>(
    `/v1/workflows/${encodeURIComponent(id)}/trigger`,
  );
  if (
    res &&
    typeof res === "object" &&
    "id" in res &&
    typeof res.id === "string"
  ) {
    return toUserWorkflow(res as UserWorkflowWire);
  }
  return getWorkflow(id);
}

/**
 * Rotate webhook secret. Plaintext `webhook_secret` is returned once.
 * Path: POST /v1/workflows/{id}/trigger/rotate-secret
 */
export async function rotateWorkflowTriggerSecret(
  id: string,
): Promise<RotateWorkflowSecretResult> {
  const res = await api.post<RotateWorkflowSecretWire & UserWorkflowWire>(
    `/v1/workflows/${encodeURIComponent(id)}/trigger/rotate-secret`,
    {},
  );
  const secret = res?.webhook_secret;
  if (!secret) {
    throw new Error("rotate-secret response missing webhook_secret");
  }
  const webhookId = res.webhook_id ?? res.trigger?.webhook_id ?? null;
  return {
    webhookSecret: secret,
    webhookUrl: relativeWebhookUrl(
      res.webhook_url ?? res.trigger?.webhook_url,
      webhookId,
    ),
    webhookId,
  };
}

/**
 * Official playbook templates (只读目录).
 * Missing route (404/501) → empty list so UI can hide the section.
 * Path: GET /v1/workflow-playbook-templates (avoid clash with /workflows/{id}).
 */
export async function listWorkflowTemplates(): Promise<WorkflowTemplate[]> {
  try {
    const res = await api.get<WorkflowTemplateWire[]>(
      "/v1/workflow-playbook-templates",
    );
    return (Array.isArray(res) ? res : [])
      .map(toWorkflowTemplate)
      .filter((t) => t.id);
  } catch (e) {
    if (isMissingRoute(e)) return [];
    throw e;
  }
}

/** Copy an official playbook into the signed-in user's workflows (使用 = 复制为我的). */
export async function createWorkflowFromPlaybook(
  input: FromPlaybookInput,
): Promise<UserWorkflow> {
  const slots: Record<string, string> = {};
  for (const [k, v] of Object.entries(input.slots)) {
    const key = k.trim();
    const val = v.trim();
    if (key && val) slots[key] = val;
  }
  const body: Record<string, unknown> = {
    playbook: input.playbook.trim(),
    slots,
  };
  const name = input.name?.trim();
  if (name) body.name = name;
  const res = await api.post<UserWorkflowWire>(
    "/v1/workflows/from-playbook",
    body,
  );
  return toUserWorkflow(res);
}

File: apps/server/agentcore/api/schemas/workflows.py
"""User workflow API schemas."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from agentcore.workflows.definition import (
    client_owned_definition,
    validate_workflow_definition,
)
from agentcore.workflows.paths import webhook_path
from agentcore.workflows.schedule import infer_schedule_preset
from agentcore.workflows.slots import (
    MAX_SLOT_DEFAULT_CHARS,
    MAX_SLOT_LABEL_CHARS,
    MAX_SLOT_VALUE_CHARS,
    MAX_SLOTS,
)
from agentcore.workflows.source import normalize_source


class WorkflowSlotModel(BaseModel):
    """One canvas slot: ``{{key}}`` in task text, ``default`` = 原轮的原值。

    Optional — a definition without slots behaves exactly as before.
    """

    key: str = Field(..., min_length=1, max_length=24)
    label: str = Field(..., min_length=1, max_length=MAX_SLOT_LABEL_CHARS)
    default: str = Field("", max_length=MAX_SLOT_DEFAULT_CHARS)

    # 声明的三个字段是给 OpenAPI / 客户端看的，不是「允许的全集」——多出来的原样带走。
    model_config = ConfigDict(extra="allow")


class WorkflowDefinitionModel(BaseModel):
    """Canvas JSON: nodes + edges (agent_step | human_gate) + optional slots.

    **校验用它，落库用** :meth:`payload` —— 不要用 ``model_dump()``。这几个字段声明的是
    「服务端认识什么」，不是「客户端能存什么」：definition 归用户所有，未知字段必须原样
    透传（``extra="allow"`` 在这里和 :class:`WorkflowSlotModel` 上都是为了这件事）。按已知
    字段重建这份 JSON，是 ``deliverable`` / ``slots`` / ``source`` 先后被抹掉的同一个根因
    —— 见 :mod:`agentcore.workflows.definition` 的所有权约定。
    """

    nodes: list[dict[str, Any]] = Field(default_factory=list)
    edges: list[dict[str, Any]] = Field(default_factory=list)
    slots: list[WorkflowSlotModel] = Field(default_factory=list, max_length=MAX_SLOTS)

    model_config = ConfigDict(extra="allow")

    @field_validator("nodes", "edges", "slots", mode="before")
    @classmethod
    def _coerce_list(cls, v: object) -> object:
        if v is None:
            return []
        return v

    @model_validator(mode="after")
    def _validate_structure(self) -> "WorkflowDefinitionModel":
        errors = validate_workflow_definition(self.payload())
        if errors:
            raise ValueError("；".join(errors))
        return self

    def payload(self) -> dict[str, Any]:
        """落库用的画布内容：客户端送来的原样内容，去掉服务端拥有的键（``source``）。"""
        return client_owned_definition(self.model_dump())


class WorkflowSourceModel(BaseModel):
    """这条工作流是从哪儿来的——服务端权威，客户端只读。

    ``kind`` 今天只有 ``"turn"``（历史行：曾经从一轮协作固化，不是现行写入），带原对话 /
    消息定位。手画的、官方模板复制的没有来源（``null``）。为什么它不在 ``definition`` 里 →
    :mod:`agentcore.workflows.source`。
    """

    kind: str
    conversation_id: str | None = None
    message_id: str | None = None

    model_config = ConfigDict(extra="allow")


class CreateWorkflowRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    description: str | None = Field(None, max_length=4000)
    definition: WorkflowDefinitionModel


class UpdateWorkflowRequest(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = Field(None, max_length=4000)
    definition: WorkflowDefinitionModel | None = None
    # Explicit clear: pass description="" → stored as NULL via route.
    clear_description: bool = False


class WorkflowTriggerModel(BaseModel):
    """One clock on a workflow: schedule XOR webhook."""

    kind: Literal["schedule", "webhook"]
    enabled: bool
    folder_id: str
    cron: str | None = None
    schedule_preset: str | None = None
    webhook_id: str | None = None
    webhook_url: str | None = None
    webhook_secret: str | None = None
    next_run_at: datetime | None = None
    last_run_at: datetime | None = None
    last_error: str | None = None


class PutWorkflowTriggerRequest(BaseModel):
    """Full replace of the workflow clock."""

    kind: Literal["schedule", "webhook"]
    folder_id: str
    enabled: bool = True
    cron: str | None = None
    schedule_preset: str | None = None

    @model_validator(mode="after")
    def _xor(self) -> "PutWorkflowTriggerRequest":
        preset = self.schedule_preset
        if self.kind == "webhook":
            if self.cron or preset:
                raise ValueError("kind=webhook 时不要传 schedule_preset / cron")
            return self
        if not self.cron and not preset:
            raise ValueError("须提供 schedule_preset 或 cron")
        if preset and preset.strip().lower() != "custom" and self.cron:
            raise ValueError("命名 schedule_preset 时不要同时传 cron")
        if preset and preset.strip().lower() == "custom" and not self.cron:
            raise ValueError("schedule_preset=custom 时须提供 cron")
        return self


class RotateWorkflowTriggerResponse(BaseModel):
    webhook_id: str
    webhook_url: str
    webhook_secret: str


class FireWorkflowWebhookResponse(BaseModel):
    conversation_id: str


class WorkflowSummary(BaseModel):
    id: str
    name: str
    description: str | None = None
    definition: dict[str, Any]
    # 与 definition 平级：客户端整份覆盖画布也带不走它。
    source: WorkflowSourceModel | None = None
    trigger: WorkflowTriggerModel | None = None
    version: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @classmethod
    def from_row(cls, row, *, webhook_secret: str | None = None) -> "WorkflowSummary":
        source = normalize_source(row.source)
        return cls(
            id=row.id,
            name=row.name,
            description=row.description,
            definition=dict(row.definition or {}),
            source=WorkflowSourceModel.model_validate(source) if source else None,
            trigger=trigger_from_row(row, webhook_secret=webhook_secret),
            version=int(row.version or 1),
            created_at=row.created_at,
            updated_at=row.updated_at,
        )


def trigger_from_row(row, *, webhook_secret: str | None = None) -> WorkflowTriggerModel | None:
    kind = getattr(row, "trigger_kind", None)
    if kind not in ("schedule", "webhook"):
        return None
    folder_id = getattr(row, "trigger_folder_id", None)
    if not folder_id:
        return None
    cron = getattr(row, "trigger_cron", None)
    webhook_id = getattr(row, "trigger_webhook_id", None)
    last_error = getattr(row, "last_trigger_error", None)
    return WorkflowTriggerModel(
        kind=kind,
        enabled=bool(getattr(row, "trigger_enabled", False)),
        folder_id=folder_id,
        cron=cron,
        schedule_preset=infer_schedule_preset(cron) if cron else None,
        webhook_id=webhook_id if kind == "webhook" else None,
        webhook_url=webhook_path(webhook_id) if webhook_id and kind == "webhook" else None,
        webhook_secret=webhook_secret,
        next_run_at=getattr(row, "trigger_next_run_at", None),
        last_run_at=getattr(row, "trigger_last_run_at", None),
        last_error=(last_error or None),
    )


class RunWorkflowRequest(BaseModel):
    folder_id: str
    # Optional per-run supplement (does not rewrite the saved definition).
    note: str | None = Field(None, max_length=16_000)
    conversation_id: str | None = None
    # Per-run slot overrides; omitted / blank keys fall back to each slot's default
    # (= the original turn's value), so an untouched form reruns the saved workflow
    # verbatim. Unknown keys are ignored — the definition owns the slot list.
    slots: dict[str, str] = Field(default_factory=dict)

    @field_validator("slots", mode="before")
    @classmethod
    def _check_slots(cls, v: object) -> object:
        if v is None:
            return {}
        if not isinstance(v, dict):
            raise ValueError("slots 必须是对象")
        if len(v) > MAX_SLOTS:
            raise ValueError(f"槽位覆盖值不能超过 {MAX_SLOTS} 个")
        for key, value in v.items():
            if not isinstance(value, str):
                raise ValueError(f"槽位 `{key}` 的值须为字符串")
            if len(value) > MAX_SLOT_VALUE_CHARS:
                raise ValueError(f"槽位 `{key}` 的值不能超过 {MAX_SLOT_VALUE_CHARS} 字")
        return v


class RunWorkflowResponse(BaseModel):
    conversation_id: str
    workflow_id: str
    workflow_version: int


class PlaybookSlotChoice(BaseModel):
    """One allowed value of an enumerated slot (render a picker, not a textbox)."""

    value: str
    label: str

    model_config = {"from_attributes": True}


class PlaybookTemplateSlotModel(BaseModel):
    """Machine-readable primary slot — clients build the copy form from this."""

    key: str
    label: str
    required: bool
    hint: str | None = None
    choices: list[PlaybookSlotChoice] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class PlaybookTemplateSummary(BaseModel):
    """Official playbook listed as a read-only workflow template."""

    id: str
    title: str
    summary: str
    # Prose one-liner for help copy; ``slots`` is the machine-readable source.
    primary_slots: str
    slots: list[PlaybookTemplateSlotModel] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class FromPlaybookRequest(BaseModel):
    """Copy an official playbook into a user workflow (use = 复制为我的)."""

    playbook: str = Field(..., min_length=1, max_length=80)
    slots: dict[str, Any] = Field(default_factory=dict)
    name: str | None = Field(None, min_length=1, max_length=200)
